File: Simulacao/MM_1/Simulacao.js
function inEvent(clientes, operador, numCliente, TR, ES, TF, HC, HS, lambda, mi) {
    TR = HC;
    clientes.push(new Cliente(numCliente, TR));
    TEC = lambda.gerarNumero();

    if (operador.ES == 0) {
        operador.changeES();
        operador.addTempoLivre(TR);

        ES = 1;
        TS = mi.gerarNumero();
        HS = TR + TS;

        clientes[0].TS = TS;
        clientes[0].HS = HS;
    }
    else {
        TF = TF + 1;
    }
    HC = TR + TEC;

    return [TR, ES, TF, HC, HS];
}

function outEvent(clientes, operador, TR, ES, TF, HC, HS, mi) {
    TR = HS;
    if (TF > 0) {
        TS = mi.gerarNumero();
        HS = TR + TS;
        TF = TF - 1;

        clientes[1].TS = TS;
        clientes[1].HS = HS;
    }
    else {
        operador.changeES();
        ES = 0;
        operador.lastTR = TR;
        HS = 999999;
    }
    return [TR, ES, TF, HC, HS];
}

function getInEvent(clientes, operador, numCliente, TR, ES, TF, HC, HS) {
    // debugger

    var lambda;
    var mi;

    if (setUp.distribuicao == 'exponencial') {
        lambda = new Exponencial(setUp.exponencial.chegada);
        mi = new Exponencial(setUp.exponencial.servico);
    }
    else if (setUp.distribuicao == 'uniforme') {
        lambda = new Uniforme(setUp.uniforme.limInfChegada, setUp.uniforme.limSupChegada);
        mi = new Uniforme(setUp.uniforme.limInfServico, setUp.uniforme.limSupServico);
    }
    else if (setUp.distribuicao == 'discreta') {
        const chegada = setUp.discreta.chegada;
        const servico = setUp.discreta.servico;
        lambda = new Discreta(chegada.valorA, chegada.probA, chegada.valorB, chegada.probB, chegada.valorC, chegada.probC);
        mi = new Discreta(servico.valorA, servico.probA, servico.valorB, servico.probB, servico.valorC, servico.probC);
    }

    return this.inEvent(clientes, operador, numCliente, TR, ES, TF, HC, HS, lambda, mi);
}

function getOutEvent(clientes, operador, TR, ES, TF, HC, HS) {
    var mi;
    if (setUp.distribuicao == 'exponencial') {
        mi = new Exponencial(setUp.exponencial.servico);
    }
    else if (setUp.distribuicao == 'uniforme') {
        mi = new Uniforme(setUp.uniforme.limInfServico, setUp.uniforme.limSupServico);
    }
    else if (setUp.distribuicao == 'discreta') {
        const servico = setUp.discreta.servico;
        mi = new Discreta(servico.valorA, servico.probA, servico.valorB, servico.probB, servico.valorC, servico.probC);
    }
    return this.outEvent(clientes, operador, TR, ES, TF, HC, HS, mi);
}

function init() {
    document.getElementById("resultTable").innerHTML = '';
    //debugger;
    var TR = 0;
    var ES = 0;
    var TF = 0;
    var HC = 0;
    var HS = 999999;
    var numCliente = 0;
    var data = [];

    var clientes = [];
    var clientesFora = [];
    var clientesFila = 0
    var operador = new Operador();

    const tempoMax = setUp.tempoMax ? setUp.tempoMax : 100;

    while (TR < tempoMax || TF != 0 || HS != 999999) {
        if (HC < HS && TR < tempoMax) {

            clientesFila += ES == 1 ? 1 : 0;

            numCliente = numCliente + 1;
            [TR, ES, TF, HC, HS] = getInEvent(clientes, operador, numCliente, TR, ES, TF, HC, HS);
            data.push(['CHEGADA', numCliente, TR, ES, TF, HC, HS]);
        }
        else {
            TF_After = TF;
            [TR, ES, TF, HC, HS] = getOutEvent(clientes, operador, TR, ES, TF, HC, HS);
            clientesFora.push(clientes.shift(0));
            data.push(['SAIDA  ', numCliente - TF_After, TR, ES, TF, HC, HS]);
        }
    }

    this.generateTable(data);
    this.getStatisticsData(clientesFora, clientesFila, numCliente, operador, TR);
    $('#separador').show();
}

function generateTable(data) {
    var inner = "";
    var linhas = [];

    data.forEach(data => {
        var str = '<tr><th>' + data[0] + '</th><td>' + data[1] + '</td><td>' + data[2] + '</td><td>' + data[3] + '</td><td>' + data[4] + '</td><td>' + data[5] + '</td><td>' + data[6] + '</td></tr>';
        linhas += str;
    });

    inner = '<div class="col-12 col-lg-12"><h4>Tabela de Simulação</h4><table class="table table-striped table-hover"><thead><tr><th scope="col">Estado</th><th scope="col">Nº Cliente</th><th scope="col">Tempo Simulacao</th><th scope="col">Estado Servidor</th><th scope="col">Tamanho Fila</th><th scope="col">Agendamento Chegada</th><th scope="col">Agendamento Saida</th></tr></thead><tbody>';
    inner += linhas;
    inner += '</tbody></table></div>';
    document.getElementById("resultTable").innerHTML = inner;
}

function getStatisticsData(clientesFora, clientesFila, numCliente, operador, TR) {
    var sumTS = 0;
    var tempoFila = 0;
    console.log(clientesFora);

    clientesFora.forEach(cliente => {
        sumTS += cliente.TS;
        tempoFila += cliente.getTempoFila();
    });
    console.log(tempoFila);
    console.log(numCliente);
    var inner = '<div class="col-12"><h4>Tabela de Resultados</h4><table class="table table-striped table-hover"><thead><tr><th scope="col">Descrição</th><th scope="col">Valor</th></tr></thead><tbody><tr><th>Tempo Médio de Espera</th><td>' + (tempoFila / numCliente).toFixed(2) + '</td></tr><tr><th>Probabilidade de Esperar</th><td>' + (clientesFila / numCliente).toFixed(2) + '</td></tr><tr><th>Probabilidade de operador livre</th><td>' + (operador.tempoLivre / TR).toFixed(2) + '</td></tr><tr><th>Tempo medio de servico</span></th><td>' + (sumTS / numCliente).toFixed(2) + '</td></tr><tr><th>Tempo medio despendido no sistema</th><td>' + (TR / numCliente).toFixed(2) + '</td></tr></tbody></table></div>';
    document.getElementById("statisticsTable").innerHTML = inner;
}

File: Simulacao/MM1_Python/Simulacao.py
from Exponencial import exponencial
from Uniforme import uniforme
from Discreta import discreta

class Cliente:
    def __init__(self, numCliente, TR):
        self.numCliente = numCliente
        self.TR = TR

    def setTS(self, TS):
        self.TS = TS
    def setHS(self, HS):
        self.HS = HS

    def getInicioServico(self):
        return self.HS - self.TS
    def getTempoFila(self):
        return self.getInicioServico() - self.TR

class Operador:
    def __init__(self):
        self.ES = 0
        self.lastTR = 0
        self.tempoLivre = 0
    
    def changeES(self):
        if self.ES == 1:
            self.ES = 0
        else:
            self.ES = 1
    def setLastTR(self, TR):
        self.lastTR = TR

    def addTempoLivre(self, TR):
        self.tempoLivre = self.tempoLivre + (TR - self.lastTR)

def inEvent(clientes, operador, numCliente, TR, ES, TF, HC, HS):
    TR = HC
    clientes.append(Cliente(numCliente, TR))
    TEC = entradaRandom.gerarNumero()
    if(operador.ES == 0):
        operador.changeES()
        operador.addTempoLivre(TR)
        ES = 1
        TS = saidaRandom.gerarNumero()
        HS = TR + TS
        clientes[0].setTS(TS)
        clientes[0].setHS(HS)
    else:
        TF = TF + 1
    HC = TR + TEC
    return TR, ES, TF, HC, HS

def outEvent(clientes, operador, TR, ES, TF, HC, HS):
    TR = HS
    if(TF > 0):
        TS = saidaRandom.gerarNumero()
        HS = TR + TS
        clientes[1].setTS(TS)
        clientes[1].setHS(HS)
        TF = TF - 1
    else:
        operador.changeES()
        ES = 0
        operador.setLastTR(TR)
        HS = 999999
    return TR, ES, TF, HC, HS

TR, ES, TF, HC = 0, 0, 0, 0
HS = 999999
numCliente = 0
entradaRandom = uniforme(4,4)
saidaRandom = uniforme(2,2)
data = []
clientes = []
clientesFora = []
operador = Operador()

sumTS = 0
clientesFila = 0
tempoFila = 0

while(TR < 10 or TF != 0 or HS != 999999):
    if(HC < HS and TR < 10):
        if ES == 1:
            clientesFila = clientesFila + 1
        numCliente = numCliente + 1
        TR, ES, TF, HC, HS = inEvent(clientes, operador, numCliente, TR, ES, TF, HC, HS)
        data.append(['CHEGADA', numCliente, TR, ES, TF, HC, HS])
    else:
        TF_After = TF
        TR, ES, TF, HC, HS = outEvent(clientes, operador, TR, ES, TF, HC, HS)
        clientesFora.append(clientes.pop(0))
        data.append(['SAIDA  ', numCliente - TF_After, TR, ES, TF, HC, HS])

print('Estado\tNumero Cliente\t\tTempo Simulacao\t\tEstado Servidor\t\tTamanho Fila\t\tAgendamento Chegada\t\tAgendamento Saida')
for [estado, _numCliente, _TR, _ES, _TF, _HC, _HS] in data:
    print(estado + '\t\t' + repr(_numCliente) + '\t\t\t' + repr(_TR) + '\t\t\t' + repr(_ES) + '\t\t\t'+ repr(_TF) + '\t\t\t\t' + repr(_HC) + '\t\t\t' + repr(_HS))


for cliente in clientesFora:
    sumTS = sumTS + cliente.TS
    #print(repr(cliente.numCliente) + ' ' + repr(cliente.TS))
    tempoFila = tempoFila + cliente.getTempoFila()


print('\nTempo medio de esperar:            %.3f' % (tempoFila/float(numCliente)))
print('Probabilidade de esperar:          %.3f' % (clientesFila/float(numCliente)))
print('Probabilidade de operador livre:   %.3f' % (operador.tempoLivre/float(TR)))
print('Tempo medio de servico:            %.3f' % (sumTS/float(numCliente)))
print('Tempo medio despendido no sistema: %.3f' % (TR/float(numCliente)))